### PR TITLE
Fix wrapping inside CLI prompts

### PR DESCRIPTION
**Pull request by ciaranmcnulty**
* Fixes long class names causing wrapping issues
* Normalises the length of progress bar blocks to match question blocks
* Adds terminal width detection so block width is dynamic
* Normalises when and where linebreaks are output

Before:
before

After:
after

### CODE REVIEW

**Comment by ciaranmcnulty** [comment]
I've made some fixes and think this is ready for re-reviewing

**Comment by ciaranmcnulty** [comment]
I want to do a 2.1.1 release, should this be in?

**Comment by stof** [comment]
This should be rebased as it conflicts with master

File: src/PhpSpec/Formatter/ConsoleFormatter.php [hunk 1]
```diff
@@ -55,48 +55,41 @@ protected function printException(ExampleEvent $event)
...
+    }
+
+    /**
+     * @param ExampleEvent $event
+     * @param string $type
+     */
+    protected function printSpecificException(ExampleEvent $event, $type)
+    {
+        $title = str_replace('\\', DIRECTORY_SEPARATOR, $event->getSpecification()->getTitle());
+//        $title = str_pad($title, $this->io->getBlockWidth(), ' ', STR_PAD_RIGHT);
```
Comment (stof) [inline src/PhpSpec/Formatter/ConsoleFormatter.php]: should be either uncommented or removed

File: src/PhpSpec/Formatter/ProgressFormatter.php [hunk 2]
```diff
@@ -81,24 +81,19 @@ function ($count) use ($total) {
...
+        $specProgress = ($stats->getTotalSpecsCount() == 0) ? 1 : ($stats->getTotalSpecs())/$stats->getTotalSpecsCount();
+        $targetWidth = ceil($this->getIO()->getBlockWidth() * $specProgress);
+        asort($counts);
 
-                return $res;
-            },
-            $percents
-        );
-        asort($barLengths);
+        $barLengths = array_map(function($count) use ($targetWidth, $counts) {
```
Comment (jon-acker) [inline src/PhpSpec/Formatter/ProgressFormatter.php]: Is this assignment necessary, (i.e. return directly)?

File: src/PhpSpec/Formatter/ProgressFormatter.php [hunk 3]
```diff
@@ -81,24 +81,19 @@ function ($count) use ($total) {
...
+    private function getBarLengths($counts)
     {
         $stats = $this->getStatisticsCollector();
-        $specProgress = ($stats->getTotalSpecs()+1)/$stats->getTotalSpecsCount();
-
-        $barLengths = array_map(
-            function ($percent) use ($specProgress) {
-                $length = $percent / 2;
-                $res = $length == 0 || $length > 1 ? floor($length * $specProgress) : 1;
+        $specProgress = ($stats->getTotalSpecsCount() == 0) ? 1 : ($stats->getTotalSpecs())/$stats->getTotalSpecsCount();
```
Comment (jon-acker) [inline src/PhpSpec/Formatter/ProgressFormatter.php]: missing space before and after / operator

File: src/PhpSpec/Formatter/ProgressFormatter.php (hunk 2 above)
Comment (ciaranmcnulty) [inline src/PhpSpec/Formatter/ProgressFormatter.php]: I think the length of the inline function would make the return harder to pick out if combined

File: src/PhpSpec/Formatter/ProgressFormatter.php (hunk 3 above)
Comment (ciaranmcnulty) [inline src/PhpSpec/Formatter/ProgressFormatter.php]: This conflicts with master so will go away

File: src/PhpSpec/Formatter/ConsoleFormatter.php (hunk 1 above)
Comment (ciaranmcnulty) [inline src/PhpSpec/Formatter/ConsoleFormatter.php]: :+1:

File: src/PhpSpec/Console/IO.php [hunk 4]
```diff
@@ -331,4 +340,27 @@ public function getBootstrapPath()
...
+     */
+    public function setConsoleWidth($width)
+    {
+        $this->consoleWidth = $width;
+    }
+
+    /**
+     * @return integer
+     */
+    public function getBlockWidth()
```
Comment (jon-acker) [inline src/PhpSpec/Console/IO.php]: spec?

Comment (ciaranmcnulty) [inline src/PhpSpec/Console/IO.php]: I don't always spec adapters

Comment (jon-acker) [inline src/PhpSpec/Console/IO.php]: but its logic. and the rest of this class seems to have specs.

Comment (ciaranmcnulty) [inline src/PhpSpec/Console/IO.php]: sigh you're probably right

**Comment by everzet** [comment]
Why are your classes so long?

**Comment by ciaranmcnulty** [comment]
@everzet Classes like `Behat\Testwork\Output\Node\EventListener\Flow\FireOnlyIfFormatterParameterListener`can get pretty long

File: src/PhpSpec/Console/IO.php (hunk 4 above)
Comment (ciaranmcnulty) [inline src/PhpSpec/Console/IO.php]: Specced

**Comment by ciaranmcnulty** [comment]
Any more comments?

**Pull request merged**